core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -273,12 +282,12 @@ class BlockManagerMasterEndpoint(
         }
       }
       bmIdsExecutor.foreach { bmId =>
-        blockManagerInfo.get(bmId).foreach { bmInfo =>
+        blockManagerInfo.get(bmId).filter(_.isAlive).foreach { bmInfo =>
```

Comment:
Does this actually have an impact ?

Reply:
Looking more, @Ngone51 do we have a leak w.r.t `blockStatusByShuffleService` in this class ?
Is bmInfo.removeBlock helping to reduce impact of the leak ?

Reply:
> Does this actually have an impact ?

Yes, this looks like a no-op as `bmIdsExecutor` should be empty for an inactive BlockManager.

> Looking more, @Ngone51 do we have a leak w.r.t blockStatusByShuffleService in this class ?
Is bmInfo.removeBlock helping to reduce impact of the leak ?

I'm not sure about it because I'm now not sure when `bmIdsExtShuffle` is not empty (or say when the block location can be external shuffle service).  @attilapiros could you give more context here?

Reply:
@Ngone51, @mridulm  So the feature was introduced by https://github.com/apache/spark/pull/24499.
It can be enabled by setting `spark.shuffle.service.fetch.rdd.enable` to true then instead of the executor's block manager ID an artificial ID will be used by where the port part is coming from the external shuffle service.

About the leak:
I cannot see leak in `blockStatusByShuffleService` regarding this method. As the clean is done a few lines above:
https://github.com/apache/spark/blob/c0a4102acbeee20abd69e1b9c98f4bc308b9caea/core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala#L279-L281

And there is test for it:
https://github.com/apache/spark/blob/c0a4102acbeee20abd69e1b9c98f4bc308b9caea/core/src/test/scala/org/apache/spark/storage/BlockManagerInfoSuite.scala#L129-L144

Reply:
@attilapiros The entry of `bmIdForShuffleService ` from `blockStatusByShuffleService` itself is not getting removed right ?
Only the corresponding blockStatus is getting removed ?

Reply:
@mridulm 

No, you are right! Thanks!

Although this is not a critical leak as there is only one `blockStatusByShuffleService` entry for each Yarn node and probably in most cases you will get a new executor allocated from an already used node. In addition the node number is much lower than the number of executors.

But for sure we should fix the small leaks too (we have a proverb for this in Hungary: "many a little becomes much"). 
I think it is fine to fix it in a separate issue (I would not block this PR) so I will create a jira issue for this.

Thanks again!

Reply:
Issue is created: https://issues.apache.org/jira/browse/SPARK-35543 

Give my some time to fix it I am very busy with another component (strange coincidence with a huge memory leak)

Reply:
Thanks for confirming @attilapiros !